### DISCUSSION TITLE
After replacing current image, Image component still shows the old image in NextJs 14

### DISCUSSION THREAD

**Discussion by hadikman** [vercel/next.js#72387] [Help] [answered]
### Summary

In a NextJs 14.2.15 project, I have a problem to showing a copied image over current image. After successful copy operation, in related page and related server component that handle of preview images files, it still shows the old image file.

For example we have */public/assets/tablet/screen.jpg* and if a user copy another image with the same filename from */public/assets/mobile/screen.jpg* to tablet directory, the copied image replaced successfully on the old file (checked by file explorer) but in the files preview, it still shows the old screen.jpg file even with use `export const dynamic = 'force-dynamic'` (in related jsx server component file) and ctrl + F5 and also "Empty Cache and Hard Reload" in browser inspect mode.
In related server action that handles copy operation after `fs.promise.copyFile()`, it runs revalidate([files preview path]) too.

I used `<img />` instead of `<Image />` with combination ctrl + F5, the problem of showing the copied file solved but I want to use native <Image /> without doing any refresh by the user to see preview of copied file.
Is there any way to disable cache <Image /> or stronger revalidate after successful copy operation?

**Comment by hadikman** [comment]
The components and server action I used:


```javascript
files.jsx

export const dynamic = 'force-dynamic'

export default function Files({ filenames }) {
  return (
    <div>
      {filenames.map(({ path, name }) => (
        <div key={name} >
            // <Image
            //   src={`/src/to/directory/${path + name}`}
            //   alt=""
            //   fill
            // />

            <img
              src={`/src/to/directory/${path + name}`}
              alt=""
            />

            // handle copy operation on client side
        </div>
      ))}
    </div>
  )
}
```

```javascript
paste-item.jsx

'use client' 

export default function PasteItem() {
  function toggleModalHandler() {
    // show modal
    // cancel operation
  }

  function replaceHandler() {
    // call pasteHandler()
  }

  function duplicateHandler() {
    // call pasteHandler()
  }

  async function pasteHandler(replaceConfirmation, duplicateConfirmation) {
    const result = await copyItemAction(
      '/public/assets/tablet/screen.jpg',
      '/public/assets/mobile/screen.jpg',
      replaceConfirmation,
      duplicateConfirmation,
    )

    if (result?.exist) {
      // show a modal to ask what to do
    } else {
      // cancel operation
    }
  }

  return (
    <div> 
      {showModal && (
        <Modal onClose={toggleModalHandler}>
          <div>
            <h3 className="mb-10">There is a file with the same name in destiantion:  what to do?
            </h3>

            <div >
              <button onClick={replaceHandler}>Replace</button>
              <button onClick={duplicateHandler}>Overwrite</button>
              <button onClick={toggleModalHandler}>Cancel</button>
            </div>
          </div>
        </Modal>
      )}
    </div>
  )
}
```

```javascript
copy-item-action.js

'use server'

import { access, copyFile } from 'fs/promises'
import path from 'path'
import { revalidatePath } from 'next/cache'
 
export default async function copyItemAction(
  source,
  destination,
  replaceConfirmation,
  duplicateConfirmation,
) {
  const parsedSrcFullPath = path.parse(srcFullPath)
  const confirmation = replaceConfirmation || duplicateConfirmation

  if (!confirmation) {
    const destFullPathWithItemName = path.join(
      destFullPath,
      parsedSrcFullPath.base,
    )
    const itemExistance = await access(destFullPathWithItemName)
      .then(() => true)
      .catch(() => false)

    if (itemExistance) return { exist: true }
  }

  destFullPath = path.join(
    destFullPath,
    duplicateConfirmation
      ? `${parsedSrcFullPath.name}-copy${parsedSrcFullPath.ext}`
      : parsedSrcFullPath.base,
  )
  try {
    await copyFile(srcFullPath, destFullPath)
    revalidatePath('/')
  } catch (err) {
    console.log(err)
  }

  return null
}
```

**Comment by jonas-is-coding** [comment, marked as answer]
The issue you are encountering is related to caching on the server side, which prevents the updated image from being displayed in your React components after the copy operation. This can happen because browsers (and Next.js by extension) cache images aggressively, even after they are replaced on the server. You are using the <img /> tag, which may not be reloading the image correctly after it’s been updated.

## Solution
### 1. Add Cache Busting to the Image URL
You can add a timestamp or a version number as a query parameter, which forces the browser to treat it as a new resource.
```javascript
export default function Files({ filenames }) {
  return (
    <div>
      {filenames.map(({ path, name }) => (
        <div key={name}>
          <img
            src={`/src/to/directory/${path + name}?v=${Date.now()}`}
            alt=""
          />
        </div>
      ))}
    </div>
  )
}
```
### 2. Apply RevalidatePath to the specific image path
You can use revalidatePath() in your server action code to revalidate the path of the copied image after the copy action. This will ensure that Next.js re-renders the page and the image is displayed correctly.
**files.jsx** (or the React Server-Component):
```javascript
export default function Files({ filenames }) {
  return (
    <div>
      {filenames.map(({ path, name }) => (
        <div key={name}>
          <img
            // Füge einen Cache-Busting-Query-Parameter hinzu
            src={`/src/to/directory/${path + name}?v=${Date.now()}`}
            alt=""
          />
        </div>
      ))}
    </div>
  )
}
```
**copy-item-action.js** (Server-Action):
```javascript
import { access, copyFile } from 'fs/promises'
import path from 'path'
import { revalidatePath } from 'next/cache'

export default async function copyItemAction(
  source,
  destination,
  replaceConfirmation,
  duplicateConfirmation,
) {
  const parsedSrcFullPath = path.parse(source)
  const confirmation = replaceConfirmation || duplicateConfirmation

  if (!confirmation) {
    const destFullPathWithItemName = path.join(
      destination,
      parsedSrcFullPath.base,
    )
    const itemExistance = await access(destFullPathWithItemName)
      .then(() => true)
      .catch(() => false)

    if (itemExistance) return { exist: true }
  }

  const destFullPath = path.join(
    destination,
    duplicateConfirmation
      ? `${parsedSrcFullPath.name}-copy${parsedSrcFullPath.ext}`
      : parsedSrcFullPath.base,
  )

  try {
    await copyFile(source, destFullPath)
    // Revalidate den spezifischen Bild-Pfad
    revalidatePath(`/src/to/directory/${parsedSrcFullPath.base}`)
  } catch (err) {
    console.log(err)
  }

  return null
}
```

## Explanation
- Cache busting: The use of ?v=${Date.now()} ensures that the browser requests a new version of the image each time without using the cache. This way you can see the newly copied file immediately.
- Revalidation: By specifically revalidating the specific image path, you ensure that Next.js re-renders the page after the change and displays the current version of the image.

**Reply by hadikman** [reply]
Thank you mate, your solution worked.

**Comment by icyJoseph** [comment]
I wonder if the problem here is that, changes to the public directory are not really allowed at runtime... 🤔 

- https://github.com/vercel/next.js/pull/17203/files

**Comment by albjeremias** [comment]
my issue was that I was showing a list of images, and the key of the list, was not index to the id of the object, but instead to the id of the "map" look, scrambling all the images 
 ```
 {objectList.map((obj, i) => (
        <Image src={obj.imgSrc} key={i}/>
      ))}
```

and it should be:
```
 {objectList.map((obj, i) => (
        <Image src={obj.imgSrc} key={obj.id}/>
      ))}
```
